Interop/Tensorflow/src/main/java/org/tribuo/interop/tensorflow/ImageConverter.java

```diff
@@ -0,0 +1,216 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.tribuo.interop.tensorflow;
+
+import com.oracle.labs.mlrg.olcut.config.Config;
+import com.oracle.labs.mlrg.olcut.config.PropertyException;
+import com.oracle.labs.mlrg.olcut.provenance.ConfiguredObjectProvenance;
+import com.oracle.labs.mlrg.olcut.provenance.impl.ConfiguredObjectProvenanceImpl;
+import org.tensorflow.ndarray.NdArrays;
+import org.tensorflow.ndarray.Shape;
+import org.tensorflow.ndarray.buffer.DataBuffers;
+import org.tensorflow.types.TFloat32;
+import org.tribuo.Example;
+import org.tribuo.Feature;
+import org.tribuo.ImmutableFeatureMap;
+import org.tribuo.Output;
+import org.tribuo.math.la.SGDVector;
+import org.tribuo.math.la.VectorTuple;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Image converter. Assumes the feature id numbers are linearised ids of the form:
+ * <pre>
+ * [0,0,0] = 0, [0,0,1] = 1, [0,1,0] = k, [1,0,0] = j*k, ..., [i,0,0] = i*j*k,
+ * </pre>
+ * That is, they are in multidimensional row major order.
+ */
+public class ImageConverter<T extends Output<T>> implements FeatureConverter<T> {
+    private static final long serialVersionUID = 1L;
+
+    @Config(mandatory=true,description="Input name.")
+    private String inputName;
+
+    @Config(mandatory=true,description="Image width.")
+    private int width;
+
+    @Config(mandatory=true,description="Image height.")
+    private int height;
+
+    @Config(mandatory=true,description="Number of channels.")
+    private int channels;
+
+    private int totalPixels;
+
+    /**
+     * For olcut.
+     */
+    private ImageConverter() {}
+
+    /**
+     * Builds an image converter for images of the supplied size.
+     * @param inputName The input name.
+     * @param width The image width.
+     * @param height The image height.
+     * @param channels The number of colour channels.
+     */
+    public ImageConverter(String inputName, int width, int height, int channels) {
+        if (width < 1 || height < 1 || channels < 1) {
+            throw new IllegalArgumentException("Inputs must be positive integers, found ["+width+","+height+","+channels+"]");
+        }
+        if (inputName == null || inputName.isEmpty()) {
+            throw new IllegalArgumentException("The input name must be a valid String");
+        }
+        long values = ((long)width)*height*channels;
+        if (values > Integer.MAX_VALUE) {
+            throw new IllegalArgumentException("Image size must be less than 2^31, found " + values);
+        }
+        this.inputName = inputName;
+        this.totalPixels = (int) values;
+        this.width = width;
+        this.height = height;
+        this.channels = channels;
+    }
+
+    /**
+     * Used by the OLCUT configuration system, and should not be called by external code.
+     */
+    @Override
+    public void postConfig() {
```

Comment:
Why is this copied, rather than called by the constructor?

Reply:
Well, it would throw `PropertyException` rather than `IllegalArgumentException` then. It didn't seem to be too much, and I'd like the validation to happen before field assignment, hence it's copied.